zebra-state/src/service/finalized_state/zebra_db/transparent.rs

```diff
@@ -125,61 +192,102 @@ impl DiskWriteBatch {
         utxos_spent_by_block: BTreeMap<OutputLocation, transparent::Utxo>,
         mut address_balances: HashMap<transparent::Address, AddressBalanceLocation>,
     ) -> Result<(), BoxError> {
-        let utxo_by_outpoint = db.cf_handle("utxo_by_outpoint").unwrap();
-        let balance_by_transparent_addr = db.cf_handle("balance_by_transparent_addr").unwrap();
+        let utxo_by_out_loc = db.cf_handle("utxo_by_outpoint").unwrap();
+        let utxo_by_transparent_addr_loc = db.cf_handle("utxo_by_transparent_addr_loc").unwrap();
 
         // Index all new transparent outputs, before deleting any we've spent
-        for (output_location, utxo) in new_outputs_by_out_loc {
-            let output = utxo.output;
-            let receiving_address = output.address(self.network());
+        for (new_output_location, utxo) in new_outputs_by_out_loc {
+            let unspent_output = utxo.output;
+            let receiving_address = unspent_output.address(self.network());
             let mut receiving_address_location = None;
 
             // Update the address balance by adding this UTXO's value
             if let Some(receiving_address) = receiving_address {
+                // TODO: fix up tests that use missing outputs,
+                //       then replace entry() with get_mut().expect()
+
                 let address_balance_location = address_balances
                     .entry(receiving_address)
-                    .or_insert_with(|| AddressBalanceLocation::new(output_location));
+                    .or_insert_with(|| AddressBalanceLocation::new(new_output_location));
                 receiving_address_location = Some(address_balance_location.address_location());
 
-                let address_balance = address_balance_location.balance_mut();
-                let new_address_balance =
-                    (*address_balance + output.value()).expect("balance overflow already checked");
+                address_balance_location
+                    .receive_output(&unspent_output)
+                    .expect("balance overflow already checked");
 
-                *address_balance = new_address_balance;
+                let address_unspent_output = AddressUnspentOutput::new(
+                    receiving_address_location.unwrap(),
+                    new_output_location,
+                );
```

Comment:
To help reading the code, feel free to improve the phrasing

```suggestion
                // Create link between the address and the new OutputLocation
                let address_unspent_output = AddressUnspentOutput::new(
                    receiving_address_location.unwrap(),
                    new_output_location,
                );
```

Reply:
Thanks for finding this complicated code without docs, I added more detailed comments in 4792c9046.